### ISSUE TITLE
kubelet-gce-e2e-ci is broken and blocking the SQ

### ISSUE THREAD

**Issue by mikedanese** [kubernetes/kubernetes#31296]
@mtaufen @Random-Liu @kubernetes/goog-node

**Comment by mikedanese** [comment]
Speculatively related to #30090

**Comment by mikedanese** [comment]
I'm not sure if a revert will fix, can someone take a look at this? https://github.com/kubernetes/kubernetes/pull/31297

**Comment by Random-Liu** [comment]
unable to create gce instance with running docker daemon for image e2e-node-ubuntu-trusty-docker9-v1-image.  googleapi: Error 403: Quota 'IN_USE_ADDRESSES' exceeded. Limit: 23.0, quotaExceeded

**Comment by Random-Liu** [comment]
Offline discussed with @yujuhong, it seems that it's still VM leak.
To unblock the submit queue, we've manually deleted some out of date VMs.

And the new build 8625 looks good till now.

**Comment by Random-Liu** [comment]
I think part of the VM leak is caused by our manually rebuild https://github.com/kubernetes/kubernetes/pull/31287#issuecomment-241876525.

In the aborted builds, the instance deletion request failed with `Error deleting instance "tmp-node-e2e-519dfbc3-coreos-alpha-1122-0-0-v20160727": googleapi: Error 404: The resource 'projects/k8s-jkns-ci-node-e2e/zones/us-central1-f/instances/tmp-node-e2e-519dfbc3-coreos-alpha-1122-0-0-v20160727' was not found, notFound`. But they are being created, it may be that the test suite tried to delete the instances before they are created, and then later the instances were successfully created, but no one would delete them again.

**Comment by lavalamp** [comment]
Update on this? It *looks* like a classic this-PR-broke-the-test situation from the jenkins dashboard. Merges have been stopped for like 5 hours now.

**Comment by yujuhong** [comment]
@lavalamp leaked VMs have already been removed. We are just waiting for the build to finish one complete run.

**Comment by mtaufen** [comment]
We also just reverted #30090 via #31297, seeing if that helps. 
Additionally, etcd was reporting a lot of fds used e.g.: `2016-08-23 21:33:41.182305 W | etcdserver: 80% of the file descriptor limit is used [used = 1021, limit = 1024]`. Might be someone is leaking fds?

**Comment by Random-Liu** [comment]
We decided to revert the PR. #30090
It seems that the newly introduced code caused file descriptor leak:
```
2016-08-23 21:32:15.503937 W | etcdserver: 80% of the file descriptor limit is used [used = 935, limit = 1024]
2016-08-23 21:32:20.504666 W | etcdserver: 80% of the file descriptor limit is used [used = 940, limit = 1024]
2016-08-23 21:32:25.505095 W | etcdserver: 80% of the file descriptor limit is used [used = 945, limit = 1024]
E0823 21:32:27.204432    1447 genericapiserver.go:743] Unable to listen for secure (open /var/run/kubernetes/apiserver.crt: no such file or directory); will try again.
2016-08-23 21:32:30.506594 W | etcdserver: 80% of the file descriptor limit is used [used = 950, limit = 1024]
2016-08-23 21:32:35.503744 W | etcdserver: 80% of the file descriptor limit is used [used = 955, limit = 1024]
2016-08-23 21:32:40.503950 W | etcdserver: 80% of the file descriptor limit is used [used = 960, limit = 1024]
E0823 21:32:42.204752    1447 genericapiserver.go:743] Unable to listen for secure (open /var/run/kubernetes/apiserver.crt: no such file or directory); will try again.
2016-08-23 21:32:45.506317 W | etcdserver: 80% of the file descriptor limit is used [used = 965, limit = 1024]
2016-08-23 21:32:50.504194 W | etcdserver: 80% of the file descriptor limit is used [used = 970, limit = 1024]
2016-08-23 21:32:55.504249 W | etcdserver: 80% of the file descriptor limit is used [used = 973, limit = 1024]
E0823 21:32:57.205060    1447 genericapiserver.go:743] Unable to listen for secure (open /var/run/kubernetes/apiserver.crt: no such file or directory); will try again.
2016-08-23 21:33:00.504235 W | etcdserver: 80% of the file descriptor limit is used [used = 978, limit = 1024]
2016-08-23 21:33:05.504604 W | etcdserver: 80% of the file descriptor limit is used [used = 983, limit = 1024]
2016-08-23 21:33:10.504905 W | etcdserver: 80% of the file descriptor limit is used [used = 988, limit = 1024]
E0823 21:33:12.205371    1447 genericapiserver.go:743] Unable to listen for secure (open /var/run/kubernetes/apiserver.crt: no such file or directory); will try again.
2016-08-23 21:33:15.504246 W | etcdserver: 80% of the file descriptor limit is used [used = 992, limit = 1024]
2016-08-23 21:33:20.503904 W | etcdserver: 80% of the file descriptor limit is used [used = 998, limit = 1024]
2016-08-23 21:33:25.503879 W | etcdserver: 80% of the file descriptor limit is used [used = 1003, limit = 1024]
E0823 21:33:27.205616    1447 genericapiserver.go:743] Unable to listen for secure (open /var/run/kubernetes/apiserver.crt: no such file or directory); will try again.
2016-08-23 21:33:30.503955 W | etcdserver: 80% of the file descriptor limit is used [used = 1008, limit = 1024]
2016-08-23 21:33:35.508235 W | etcdserver: 80% of the file descriptor limit is used [used = 1013, limit = 1024]
2016-08-23 21:33:40.503934 W | etcdserver: 80% of the file descriptor limit is used [used = 1018, limit = 1024]
```
Then apiserver is gone. We are debugging this now.

**Comment by mtaufen** [comment]
Looks like a new fd is consumed every second.

**Comment by lavalamp** [comment]
1024 FD limit is too low. kube-apiserver requires much more. (I would have said it may not matter for 1 client but it looks like it does)

**Comment by lavalamp** [comment]
Actually after chatting with Lantao it sounds like maybe connections are
not getting reused, which would trigger this.

**Comment by Random-Liu** [comment]
Thanks @lavalamp, that is the problem! But not on apiserver, the problem is in node e2e test framework.

@mtaufen and I found that in node e2e test framework, we use `http.Get` to do health check https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/e2e_service.go#L470.
However, we only check whether it responds 200, never read the content returned. This will make it create a new connection every time we call it.

It was fine before #30090, because we only do readiness check for several seconds after we start services. However, now we run readiness check during the whole test to make sure kubelet is running, which leaks a lot for file descriptors.

**Comment by ronnielai** [comment]
Using http.Head in https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/e2e_service.go#L470 should avoid the leak.

**Comment by Random-Liu** [comment]
Offline discussed with @dchen1107, the kubelet-gce-e2e-ci is green now after reverted #30090.
Close this issue, and will file another issue to track https://github.com/kubernetes/kubernetes/issues/31296#issuecomment-241912256.

**Comment by mikedanese** [comment]
This is very broken again.

**Comment by mikedanese** [comment]
It seems like sometimes a set of VMs are created and never deleted. There seem to be a few VMs that are lingering from old test runs.

**Comment by Random-Liu** [comment]
@mikedanese I'll take a look.

**Comment by mikedanese** [comment]
https://storage.googleapis.com/kubernetes-jenkins/logs/kubelet-gce-e2e-ci/8634/build-log.txt
https://storage.googleapis.com/kubernetes-jenkins/logs/kubelet-gce-e2e-ci/8635/build-log.txt
https://storage.googleapis.com/kubernetes-jenkins/logs/kubelet-gce-e2e-ci/8636/build-log.txt

**Comment by Random-Liu** [comment]
It is caused by VM leak.
The VMs are leaked by `kubelet-serial-gce-e2e-ci`. The reason is that the test is aborted because of 90 minutes timeout. And the instances were left there.
```
19:17:13 I0823 19:17:13.795887    6239 run_e2e.go:450] Creating instance {image:e2e-node-ubuntu-trusty-docker9-v1-image project:kubernetes-node-e2e-images metadata:<nil> machine: tests:[]}
19:22:58 I0823 19:22:58.051574    6239 e2e_remote.go:224] Starting tests on tmp-node-e2e-def97ee9-e2e-node-ubuntu-trusty-docker9-v1-image
20:43:48 Build timed out (after 90 minutes). Marking the build as failed.
20:43:48 Build was aborted
```
And the instance is left there:
```
tmp-node-e2e-def97ee9-e2e-node-ubuntu-trusty-docker9-v1-image	us-central1-f	2016-08-23T19:17:14.594-07:00
```

**Comment by Random-Liu** [comment]
We should:
* Make run_e2e.go be able to handle abort.
* Extend the jenkins timeout for serial test.

**Comment by mikedanese** [comment]
The emergency timeout increase SGTM but we should cleanup old test runs at the beginning of the test and revert the timeout increase. This test is failing again.

**Comment by Random-Liu** [comment]
@mikedanese The timeout increase is necessary, the test takes quite a lot of time.

**Comment by mikedanese** [comment]
Ok, I restarted the test, merged the timeout increase, and cleaned up the all vms.

**Comment by mikedanese** [comment]
> The timeout increase is necessary, the test takes quite a lot of time.

Yes but maybe not a 90 -> 600 minute timeout increase :)

Can we also clean the vms up at the beginning of test runs like we do in other e2es?

**Comment by dchen1107** [comment]
@mikedanese yes, node-e2e should do that, but that is not how it currently designed & implemented. Some of tests are recreating the node, others are just re-using the existing one. We should fix this.

**Comment by Random-Liu** [comment]
> Can we also clean the vms up at the beginning of test runs like we do in other e2es?

@mikedanese We already have periodically cleanup https://github.com/kubernetes/test-infra/pull/52, but the VMs are leaking too quickly this time... :( Let's see whether we can find some way to do some cleanup at the beginning of the test. :)

> Yes but maybe not a 90 -> 600 minute timeout increase :)

A lot of test suites are using the `jenkins-timeout` (600 minute) https://github.com/kubernetes/test-infra/blob/9a90753030eb511b1c5bd29c6c4b74952b914f7b/jenkins/job-configs/global.yaml#L70-L71. Because the test itself has a 3 hour timeout https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/jenkins/jenkins-serial.properties#L12, so it should be safe to set 600 minute on jenkins side. :)
